src/server/string_family.cc

```diff
@@ -1105,17 +1118,27 @@ void StringFamily::SetExGeneric(bool seconds, CmdArgList args, ConnectionContext
   int64_t unit_vals;
 
   if (!absl::SimpleAtoi(ex, &unit_vals)) {
-    return cntx->SendError(kInvalidIntErr);
+    return cntx->SendError(kInvalidIntErr, kSyntaxErrType);
   }
 
   if (unit_vals < 1) {
     return cntx->SendError(InvalidExpireTime(cntx->cid->name()));
   }
 
+  DbSlice::ExpireParams expiry{
+      .value = unit_vals,
+      .unit = seconds ? TimeUnit::SEC : TimeUnit::MSEC,
+      .absolute = false,
+  };
+
+  int64_t now_ms = GetCurrentTimeMs();
+  auto [rel_ms, abs_ms] = expiry.Calculate(now_ms, false);
+  if (abs_ms < 0)
+    return cntx->SendError(InvalidExpireTime("set"));
+
   SetCmd::SetParams sparams;
   sparams.flags |= SetCmd::SET_EXPIRE_AFTER_MS;
-  sparams.expire_after_ms =
-      DbSlice::ExpireParams::Cap(unit_vals * (seconds ? 1000 : 1), TimeUnit::MSEC);
+  sparams.expire_after_ms = expiry.Calculate(now_ms, true).first;
```

Comment:
why don't you re-use rel_ms at the bottom? Because you don't cap during the fist call... but why? I'm in confusion 🙂

Reply:
if I cap, I "fix" the potential overflow, so it won't return a negative `abs_ms`. The right thing would be to remove `cap` parameter and do it separately, via a different call but I it's not worth the time